### ISSUE TITLE
bot crash

### ISSUE THREAD

**Issue by rizar** [mila-udem/babyai#7]
When I am generating demonstrations for the BossLevel, I get this:

```
2019-01-04 14:28:54,300: ERROR: error while generating demo #325
Traceback (most recent call last):
  File "scripts/make_agent_demos.py", line 96, in generate_demos
    action = agent.act(obs)['action']
  File "/home/dzmitry/Dist/baby-ai-game/babyai/utils/agent.py", line 150, in act
    action = self.bot.get_action()
  File "/home/dzmitry/Dist/baby-ai-game/babyai/bot.py", line 1247, in get_action
    action = subgoal.get_action()
  File "/home/dzmitry/Dist/baby-ai-game/babyai/bot.py", line 478, in get_action
    return self.subgoal_accomplished()
  File "/home/dzmitry/Dist/baby-ai-game/babyai/bot.py", line 69, in subgoal_accomplished
    return subgoal.get_action()
  File "/home/dzmitry/Dist/baby-ai-game/babyai/bot.py", line 329, in get_action
    return self.subgoal_accomplished()
  File "/home/dzmitry/Dist/baby-ai-game/babyai/bot.py", line 69, in subgoal_accomplished
    return subgoal.get_action()
  File "/home/dzmitry/Dist/baby-ai-game/babyai/bot.py", line 194, in get_action
    obj_pos = self.bot.find_obj_pos(key_desc)
  File "/home/dzmitry/Dist/baby-ai-game/babyai/bot.py", line 985, in find_obj_pos
    assert len(obj_desc.obj_set) > 0
AssertionError
```

Unless this stack trace is immediately meaningful for anyone, I'm going to dig into it myself.

**Comment by saleml** [comment]
It seems like the instruction makes the bot try to look for a particular key, but there is no corresponding key in the environment.
Do you happen to know the corresponding seed ?

**Comment by rizar** [comment]
I don't have the seed, but I can tell you that this is the demonstration #325.

**Comment by rizar** [comment]
FYI, I'm debugging this now.

**Comment by rizar** [comment]
Ok, so here is what happens. The instruction is `go to the yellow key and go to the grey door, then go to a box and open the green door`. But the yellow key happens to be behind the green door. So the green key is picked up and the green door is opened before the agent proceeds to executing the last clause. 

Will keep investigating.

**Comment by rizar** [comment]
@maximecb I might be wrong, but it seems to me that this instruction is basically impossible to execute. Because the `open the green door` clause can only be executed if the green door is originally closed. Am I right?

**Comment by maximecb** [comment]
For unlocked doors it's possible to close them and then reopen them, and I think the bot already does this. For locked doors, I believe you're right, it's not possible to close and then reopen them. This once again goes to show that our seemingly simple gridworld has some complex emergent interactions that are hard to foresee.

Some possibilities to handle this:
1. For levels with implicit unlocking, reject instructions that require opening the locked door.
2. Weaker form of 1, reject instructions requiring opening the locked door, unless this is the only instruction, or unless there is no sequencing, eg (`go to the grey box and open the green door`) should be ok.
3. Reject instructions that require opening the locked door and also mention objects in the locked room.

**Comment by rizar** [comment]
> When a locked door is opened, replace it by an unlocked door so that it can be closed and opened again (without requiring a key). We could make this change in MiniGrid.

That sounds like the "real world" behavior of doors, doesn't it? :)

Just to make sure I understand: is it currently the case that if a locked door is opened and then closed then it becomes locked it, but the key disappears from the world?

**Comment by maximecb** [comment]
Right now, you can open and close "normal" (unlocked) doors as many times as you want. However, after you open a locked door, you can't close it back. This is because doors and locked doors are different types of objects in the world, which in hindsight was a bad design decision.

The cleanest fix would be to try to "unify" doors and locked doors into one type of object, so that doors have 3 states (locked, closed and open). Opening a locked door would transition to the open state, which can only transition to the closed state, but not to locked again.

**Comment by rizar** [comment]
From the top of my head, the fix that you propose seems to make a lot of
sense. If you are not too afraid of backward compatibility issues, I guess
we could go ahead with it.

**Comment by maximecb** [comment]
I'll be looking at it tomorrow. Keep in mind you'll have to regenerate all the demos after that change (assuming it works). Still good to get all your scripts tested and ready though.

**Comment by rizar** [comment]
Of course, I am not going to launch any large scale generation before we
are done with this. OTOH, my scripts are fully ready.

**Comment by maximecb** [comment]
The change is merged and the automated tests passed. Visual inspection looks right as well. Fingers crossed that the demos will generate correctly now.

**Comment by rizar** [comment]
OK, thanks, now it crashes on the example #709 with a Python stack overflow.  The instruction is "open purple door". I will investigate.


![image](https://user-images.githubusercontent.com/654434/50850248-4a336a00-1347-11e9-8101-648f848dd6ce.png)

**Comment by rizar** [comment]
See a proposed fix in #10

**Comment by rizar** [comment]
Looks like the bot no longer crashes
